txscript/script.go

```diff
@@ -151,6 +169,46 @@ func hasP2SHRedeemScriptStakeOpCodes(version uint16, sigScript, pkScript []byte)
 	return nil
 }
 
+// hasP2SHRedeemScriptTreasuryOpCodes returns an error if the provided public key
+// script is a regular pay-to-script-hash or a stake-tagged pay-to-script and,
+// when it is, that the redeem script within the provided signature script
+// contains stake opcodes.  An error is also returned if the signature script is
+// malformed after determining the public key script is one of the
+// aforementioned cases.
+func hasP2SHRedeemScriptTreasuryOpCodes(version uint16, sigScript, pkScript []byte) error {
+	// The only stake scripts currently supported are version 0.
+	if version != 0 {
+		return nil
+	}
+
+	// Nothing further to check if the public key script is not a normal
+	// pay-to-script-hash script or one tagged with a stake opcode.
+	if !(isScriptHashScript(pkScript) || isTreasuryScriptHashScript(pkScript)) {
```

Comment:
This is certainly incorrect.  It would allow treasury opcodes in stake-tagged script hash scripts.  The reverse is also true and this code is incorrect for it to.   That is to say the existing stake opcodes must not be allowed in treasury-tagged script hash scripts when the `ScriptVerifyTreasury` flag is set.

I think the better way to handle this is to consider that the treasury opcodes are also stake opcodes and treasury-tagged script hash scripts when `ScriptVerifyTreasury` is set and to work it into the existing functions accordingly.

Reply:
I think I fixed this. Please have another look.